### PR TITLE
DT-2479: ✨ Start to add in retrieval of linked account information

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/UserSelectorAndMfaAuthorizationEndpoint.kt [hunk 1]
```diff
@@ -111,10 +111,10 @@ class UserSelectorAndMfaAuthorizationEndpoint(
   ): Authentication? {
     val source = account.substringBefore("/")
     val username = account.substringAfter("/")
-    val azureUser = authentication.principal as UserDetailsImpl
+    val loginUser = authentication.principal as UserDetailsImpl
 
     val user = userService.getMasterUserPersonDetailsWithEmailCheck(
-      username, AuthSource.fromNullableString(source), azureUser.userId
+      username, AuthSource.fromNullableString(source), loginUser
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/UserSelectorAndMfaAuthorizationEndpoint.kt:117]: have to pass the user instead as only azure users have their email address as their `userId`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserContextApprovalHandler.kt [hunk 2]
```diff
@@ -14,6 +14,7 @@ class UserContextApprovalHandler(
   private val userContextService: UserContextService,
   clientDetailsService: ClientDetailsService,
   private val mfaClientService: MfaClientService,
+  private val linkAccounts: Boolean,
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserContextApprovalHandler.kt:17]: added for use in next PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserService.kt [hunk 3]
```diff
@@ -78,20 +79,20 @@ class UserService(
...
-      none -> false
-    }
+  fun getEmail(userPersonDetails: UserPersonDetails): String? {
+    // special case where the authentication is passed through - we won't have the email address in that case
+    val email = if (userPersonDetails is UserDetailsImpl || userPersonDetails.authSource == nomis.source) {
+      verifyEmailService.getEmail(userPersonDetails.username).filter { it.verified }.map { it.email }.orElse(null)
+    } else with(userPersonDetails.toUser()) { if (verified) email else null }
+
+    return if (userPersonDetails.authSource == nomis.source) {
+      email ?: getEmailAddressFromNomis(userPersonDetails.username).orElse(null)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserService.kt:89]: fallback to check nomis if we don't find a verified email in auth for nomis users

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserService.kt [hunk 4]
```diff
@@ -122,10 +123,10 @@ class UserService(
 
   fun getEmailAddressFromNomis(username: String): Optional<String> {
     val emailAddresses = verifyEmailService.getExistingEmailAddressesForUsername(username)
-    val justiceEmail = emailAddresses
-      .filter(UserService::isHmppsEmail)
+    val notHmpsGsiEmails = emailAddresses
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserService.kt:126]: this was odd - there are a variety of email addresses in NOMIS that are allowed and aren't justice email address, so we should copy them through too.  It is just the old `hmps.gsi.gov.uk` that we don't want anymore.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/integration/ClientLoginSpecification.kt [hunk 5]
```diff
@@ -151,16 +151,16 @@ class ClientLoginSpecification : AbstractDeliusAuthSpecification() {
   @Test
   fun `I can sign in from another client as azure ad user with a nomis account`() {
     // The email is mapped to RO_USER in the nomis database
-    azureClientSignIn("phillips@fredjustice.gov.uk")
-      .jsonPath(".user_name").isEqualTo("RO_USER")
-      .jsonPath(".user_id").isEqualTo("4")
-      .jsonPath(".sub").isEqualTo("RO_USER")
+    azureClientSignIn("ro_user@some.justice.gov.uk")
+      .jsonPath(".user_name").isEqualTo("RO_DEMO")
```
Comment (petergphillips) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/integration/ClientLoginSpecification.kt:155]: had to switch to a different user as `RO_USER` had two justice.gov.uk accounts in nomis and could have been either of them.  We now only allow one email address for a user (ignoring any hmps.gsi.gov.uk emails).

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**